File: python/leetcode/054.py
class Solution:
    def spiralOrder(self, matrix):
        if len(matrix) < 1: return []
        m, n = len(matrix), len(matrix[0])
        k, l = 0, 0
        res = []
        while k < m and l < n:
            for i in range(l, n): res.append(matrix[k][i])
            k += 1
            for i in range(k, m): res.append(matrix[i][n-1])
            n -= 1
            if k < m:
                for i in range(n-1, l-1, -1): res.append(matrix[m-1][i])
                m -= 1
            if l < n:
                for i in range(m-1, k-1, -1): res.append(matrix[i][l])
                l += 1
        return res

File: cpp/leetcode/054.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>> &matrix) {
        if (matrix.empty() || matrix[0].empty()) {
            return {};
        }
        vector<int> res;
        int m = matrix.size(), n = matrix[0].size();
        int left = 0, top = 0;
        while (top < m && left < n) {
            for (int i = left; i < n; ++i) {
                res.push_back(matrix[top][i]);
            }
            top++;
            for (int i = top; i < m; ++i) {
                res.push_back(matrix[i][n - 1]);
            }
            n--;
            if (top < m) {
                for (int i = n - 1; i > left - 1; --i) {
                    res.push_back(matrix[m - 1][i]);
                }
                m--;
            }
            if (left < n) {
                for (int i = m - 1; i > top - 1; --i) {
                    res.push_back(matrix[i][left]);
                }
                left++;
            }
        }
        return res;
    }
};